diksha/diksha.init.js

```diff
@@ -52,11 +52,22 @@ let initializeDikshaData = (path) => {
 
 let processEcarFiles = (filePath) => {
     return readdir(filePath)
-        .then(files => {	
+        .then(files => {
             return files
                 .filter(file => file.endsWith('.ecar'))
                 .reduce((chainedExtractPromises, file) => {
-                    return chainedExtractPromises.then(() => extractFile(filePath, file));
+                    return chainedExtractPromises.then(
+                        () => {
+                            console.log(`Extraction of ${file} completed.`);
+
+                            return extractFile(filePath, file);
+                        },
+                        err => {
```

Comment:
it's never a good idea to provide `err` as the second callback, `err` should always be cached in the catch block. 
second callback always gets triggered for previous thens, but in your case you don't have a previous then so technically your error will never get caught and whole process will crash... please refer to this
https://stackoverflow.com/questions/33278280/promise-then-vs-then-catch

Reply:
@rak16

Reply:
Good catch, will take care of this..

Reply:
I thought it through and  I am not really sure how this would fail. Let me try to explain it from my perspective and then you can let me know where does it exactly fails, if at all. I used [this article](https://css-tricks.com/why-using-reduce-to-sequentially-resolve-promises-works/) as a reference and tweaked it a little to prevent fail-fast.

So after `reduce`, the promise structure would look something like,

```
q.when()
    .then(a, b)
    .then(a, b)
    ....
    .catch(b) // This catch is present where the function is being called from
```

When I look at it this way, there is an error handler for every `a` finally terminating at a catch which takes care of the last one. Thus, all the errors are caught immediately after they fail and control keeps flowing.

Let me know if there's something I am missing.

Reply:
@rak16 's logic looks good to me, as each error is either caught at the next `then` or the final `catch`. However, I would prefer if *all* error catches were local to the function, instead of risking indeterministic behavior due to varying callers.